README.md

```diff
@@ -120,6 +120,73 @@ import utc from '@momentjs/moment/timezones/utc/'; // Note trailing slash
 // Error: folders cannot be imported (there is no index.* magic)
 ```
 
+### 2. Imports Field
+
+Imports provide the ability to remap bare specifiers within packages before they hit the node_modules resolution process.
+
+The current proposal prefixes all imports with `#` to provide a clear signal that it's a _symbolic specifier_ and also to prevent packages that use imports from working in any environment (runtime, bundler) that isn't aware of imports.
```

Comment:
I like `~` for this instead of `#`. Parcel and a few other tools already support `~/foo` to mean `foo` within the nearest folder with package.json. Supporting `~foo` (without the slash) to mean the `foo` named import within the package.json kinda makes sense too. `~` always refers to the folder with package.json, and you can either refer to a file or a named import from there. Not sure if Node is interested in the `~/` specifier as well, but it would leave the option open for the future.

Reply:
I think node may be interested in `~/` (or something like it) to mean "this package as exported". Right now there's no good way to unit test the public interface when using `exports` for example. If not `~`, we'd need to find another character for this.

Reply:
I guess that's what I meant. `~/` gets you to the package root, and it's normal resolution after that. So if there is a `foo` export, then `~/foo` would refer to that as it would normally. And `~foo` would refer to an import. `~` or `~/` by itself could refer to `main`.

Reply:
Is it worth opening a dedicated issue to resolve the sigil question? My current thoughts:

We have three kinds of specifiers that people have asked for (implied: that we may or may not want to support):

1. Getting the public interface of the package. This will allow people using exports to actually unit test their packages (since `exports` do not apply to relative specifiers). Examples: `✩` (the main/default export), `✩/subpath`
2. Adding custom aliases that are only valid inside of the package boundary. Examples: `✩data/emoji.json`, `✩fetch`.
3. Accessing arbitrary paths relative to the package boundary. Examples: `✩/src/model.mjs`.

Of these, only (1) and (3) actually conflict. (2) could share a symbol with either one of them. A concern raised by @guybedford was that if (1) and (2) share a symbol, it may be confusing.

So to me the options are:

1. One sigil, no (3):
  1. Use `~` and `~/` to mean "this package as if it was imported by name".
  1. Allow `~<name>` to be used for custom aliases within the package.
2. Two sigils, potentially support for (3) as well:
  1. Use `~` and ~/` to mean "this package as if it was imported by name".
  1. Use `#<name>` for "private names", aliases only visible inside of the package.
  1. (optional) Use `#/` for "paths relative to the package boundary".

For packages that don't use `exports`, `~/` and `#/` are effectively the same but that would change once they choose to remap subpaths and/or lock themselves down.

Reply:
My preference right now would be `~` + `#<name>` without support for importing non-public paths relative to the project root. There would still be design space for adding `#/` in the future if it becomes truly necessary.

Reply:
Isn't `#` problematic since it's meaningful to URL parsing? ESM specifiers are URLs, so wouldn't that be considered a hash?

Reply:
While ESM uses URLs as cache keys in the browser, it has more restrictions for specifiers. The only kinds of specifiers it allows are:

1. Relative specifiers starting with `./`, `../`, or `/`.
2. An absolute URL, including protocol.

See: https://html.spec.whatwg.org/#resolve-a-module-specifier

So neither `?` nor `#` may start a specifier unless something like an import map is involved. Even though both are valid relative URLs in other contexts like certain HTML attributes.

Reply:
I thought we would promise not to bikeshed yet :P

But if we must then my preference would be to use `~/` for the internal root and `#/` or something else for the public interface.

Under that logic, perhaps we should use `~name`?

But yeah opening a new issue to hash / tilde this out seems to make sense!

Reply:
Forked this to an issue: https://github.com/jkrems/proposal-pkg-exports/issues/41